### PR TITLE
Implement ControlState Instance that writes "upgrade in progress (next checkpoint)" to reserved pages

### PR DESCRIPTION
**Pull request by yontyon**

### CODE REVIEW

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 1]
```diff
@@ -14,12 +14,32 @@
...
+  std::stringstream inStream;
+  inStream.str(str);
+  controlStateMessages::StopAtNextCheckpointMessage msg;
+  concord::serialize::Serializable::deserialize(inStream, msg);
+  return msg.stop_;
+}
 
 ControlStateManager::ControlStateManager(IStateTransfer& state_transfer) : state_transfer_{state_transfer} {
-  state_transfer_.getStatus();  // Temporary, to escape not-used compilation error
+  scratchPage_ = (char*)std::malloc(size_of_reserved_pages_);
```
Comment (f-squirrel) [inline bftengine/src/bftengine/ControlStateManager.cpp:42]: Do we really need manual memory management?
The same effect can be achieved by using a regular `std::string`.
1. It has out-of-the-box integration with `std::stringsream`.
2. We can remove `malloc` and `free`.
3. The dtor can be `=default`.

In the constructor the size of the string can be set like this: `scratchPage_.resize(size_of_reserved_pages_);`
Once the string has to be passed to the `serialize` functions, we may pass `scratchPage_.data()`.
Note, `data()` returns non-constant pointer to char.

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 2]
```diff
@@ -14,12 +14,32 @@
 #include "ControlStateManager.hpp"
 namespace bftEngine {
 
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint() {
+  std::stringstream outStream;
```
Comment (f-squirrel) [inline bftengine/src/bftengine/ControlStateManager.cpp:20]: Nit: `std:ostringstream`?

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 3]
```diff
@@ -14,12 +14,32 @@
...
+}
 
-std::optional<uint64_t> ControlStateManager::getStopCheckpointToStopAt() { return UINT64_MAX; }
+std::optional<bool> ControlStateManager::getStopCheckpointToStopAt() {
+  if (!state_transfer_.loadReservedPage(
+          resPageOffset() + reserved_pages_indexer_.update_reserved_page_, size_of_reserved_pages_, scratchPage_)) {
+    return {};
+  }
+  std::string str(scratchPage_, size_of_reserved_pages_);
+  std::stringstream inStream;
```
Comment (f-squirrel) [inline bftengine/src/bftengine/ControlStateManager.cpp:34]: Nit: `std:istringstream`?

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 4]
```diff
@@ -14,19 +14,46 @@
...
 
   ControlStateManager(IStateTransfer& state_transfer);
   ControlStateManager& operator=(const ControlStateManager&) = delete;
   ControlStateManager(const ControlStateManager&) = delete;
-
+  ~ControlStateManager() { free(scratchPage_); }
  private:
   IStateTransfer& state_transfer_;
+
+  const uint32_t size_of_reserved_pages_ = 512;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: done

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 5]
```diff
@@ -14,12 +14,32 @@
...
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint() {
+  std::stringstream outStream;
+  controlStateMessages::StopAtNextCheckpointMessage msg(true);
+  concord::serialize::Serializable::serialize(outStream, msg);
+  auto data = outStream.str();
+  state_transfer_.saveReservedPage(
+      resPageOffset() + reserved_pages_indexer_.update_reserved_page_, data.size(), data.c_str());
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: done

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 1 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: done

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 2 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: done

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 3 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: done

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 6]
```diff
@@ -14,12 +14,32 @@
...
 
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint() {
+  std::ostringstream outStream;
+  controlStateMessages::StopAtNextCheckpointMessage msg(true);
+  concord::serialize::Serializable::serialize(outStream, msg);
+  auto data = outStream.str();
+  state_transfer_.saveReservedPage(
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ControlStateManager.cpp:24]: Not related to this PR. Please change state_transfer_ to pointer.

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 7]
```diff
@@ -14,19 +14,47 @@
...
 
   ControlStateManager(IStateTransfer& state_transfer);
   ControlStateManager& operator=(const ControlStateManager&) = delete;
   ControlStateManager(const ControlStateManager&) = delete;
+  ~ControlStateManager() {}
 
  private:
   IStateTransfer& state_transfer_;
+
+  static constexpr uint32_t size_of_reserved_pages_ = 512;
```
Comment (toly-kournik) [inline bftengine/include/bftengine/ControlStateManager.hpp:35]: Size of reserved page is defined in configuration and can't be changed by a module

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 8]
```diff
@@ -14,19 +14,47 @@
...
+    uint32_t update_reserved_page_ = 0;
+  };
+
+  reservedPageIndexer reserved_pages_indexer_;
+};
+
+namespace controlStateMessages {
+
+// control state messages
+class StopAtNextCheckpointMessage : public concord::serialize::SerializableFactory<StopAtNextCheckpointMessage> {
```
Comment (toly-kournik) [inline bftengine/include/bftengine/ControlStateManager.hpp:49]: TODO @toly-kournik : subclass from Serializable

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 7 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp:35]: The size is unique for all the reserved pages users?

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 6 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp:24]: It is a reference, why does it matter if this is a pointer or reference?

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 9]
```diff
@@ -14,19 +14,47 @@
 #pragma once
 #include <optional>
 #include "IStateTransfer.hpp"
+#include "ReservedPages.hpp"
+#include "Serializable.h"
 
 namespace bftEngine {
 
-class ControlStateManager {
+class ControlStateManager : public ResPagesClient<ControlStateManager, 1, 1> {
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:22]: Nit: Without looking at the definition of `ResPagesClient`, I have no idea what these parameters mean. Maybe create some named constexprs for the `1s` to make it a bit clearer to the reader.

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 10]
```diff
@@ -14,19 +14,47 @@
...
  public:
   void setStopAtNextCheckpoint();
-  std::optional<uint64_t> getStopCheckpointToStopAt();
+  std::optional<bool> getStopCheckpointToStopAt();
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:25]: Why not return the actual checkpoint?

Comment (andrewjstone) [reply bftengine/include/bftengine/ControlStateManager.hpp:25]: Nit: Naming. `getCheckpointToStopAt` is more legible.

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 7 above)
Comment (andrewjstone) [reply bftengine/include/bftengine/ControlStateManager.hpp:35]: Not directly related to this PR, but to the resPagesClients themselves. What happens if the number of reserved pages used by all clients exceeds the configuration? Do we have any detection of this? @toly-kournik

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 8 above)
Comment (andrewjstone) [reply bftengine/include/bftengine/ControlStateManager.hpp:49]: Once #776 goes in I think we should use concord message format for these messages. Serializable is fine for local persistence, but it doesn't take into account endianness and is tightly coupled  to C++ type names. That should not hold up merging this though. The change can be made in a future commit.

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 6 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: fixed

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 7 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: fixed

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 9 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: fixed

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 10 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: fixed

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 11]
```diff
@@ -14,19 +14,53 @@
...
 
-  ControlStateManager(IStateTransfer& state_transfer);
+  ControlStateManager(IStateTransfer* state_transfer, uint32_t sizeOfReservedPages);
   ControlStateManager& operator=(const ControlStateManager&) = delete;
   ControlStateManager(const ControlStateManager&) = delete;
+  ~ControlStateManager() {}
 
  private:
-  IStateTransfer& state_transfer_;
+  IStateTransfer* state_transfer_;
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:38]: What was the rationale for switching to a pointer? I never saw a response from @toly-kournik on this.

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 12]
```diff
@@ -14,12 +14,34 @@
 #include "ControlStateManager.hpp"
 namespace bftEngine {
 
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ControlStateManager.cpp:17]: "does not aware to" -> "is not aware of"

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 13]
```diff
@@ -14,12 +14,34 @@
...
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint(int64_t currentSeqNum) {
+  uint64_t seq_num_to_stop_at = (currentSeqNum + 2 * checkpointWindowSize);
+  seq_num_to_stop_at = seq_num_to_stop_at - (seq_num_to_stop_at % checkpointWindowSize);
+  std::ostringstream outStream;
+  controlStateMessages::StopAtNextCheckpointMessage msg(seq_num_to_stop_at);
+  concord::serialize::Serializable::serialize(outStream, msg);
+  auto data = outStream.str();
+  state_transfer_->saveReservedPage(
+      resPageOffset() + reserved_pages_indexer_.update_reserved_page_, data.size(), data.data());
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ControlStateManager.cpp:27]: Can we please encapsulate `resPageOffset() + reserved_pages_indexer_.update_reserved_page_` in a method and explain what it's doing with a header comment if necessary.

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 14]
```diff
@@ -14,19 +14,53 @@
...
+  const uint32_t sizeOfReservedPage_;
+  std::string scratchPage_;
+
+  // reserved page indexer
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:42]: This comment doesn't tell me anything more than the variable name. A better comment would explain why the reservedPageIndexer is needed and what it is used for.

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 15]
```diff
@@ -14,19 +14,53 @@
...
+static constexpr uint32_t ControlHandlerStateManagerReservedPagesIndex = 1;
+static constexpr uint32_t ControlHandlerStateManagerNumOfReservedPages = 1;
+class ControlStateManager : public ResPagesClient<ControlStateManager,
+                                                  ControlHandlerStateManagerReservedPagesIndex,
+                                                  ControlHandlerStateManagerNumOfReservedPages> {
  public:
-  void setStopAtNextCheckpoint();
-  std::optional<uint64_t> getStopCheckpointToStopAt();
+  void setStopAtNextCheckpoint(int64_t currentSeqNum);
+  std::optional<int64_t> getCheckpointToStopAt();
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:30]: We use uint64_t below, but int64_t here. Why?

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 16]
```diff
@@ -14,19 +14,53 @@
...
+  };
+
+  reservedPageIndexer reserved_pages_indexer_;
+};
+
+namespace controlStateMessages {
+
+// control state messages
+class StopAtNextCheckpointMessage : public concord::serialize::SerializableFactory<StopAtNextCheckpointMessage> {
+  const std::string getVersion() const override { return "1"; }
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:54]: Why a string vs an int?

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 11 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp:38]: All state transfer holders, are holding it via a raw pointer (for now).
@toly-kournik wants it to be the same in all.

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 16 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp:54]: This is the interface of the concord::serialize

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 15 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp:30]: SeqNum is a typedef for int64_t. I will change the other one to be int64_t

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 12 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: fixed

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 13 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: done

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 14 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: done

**Pull request merged**